File: javascript/scrape_to_markdown.js
#!/usr/bin/env node
// scrape_to_markdown.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/web-markdown-scraper/scripts/scrape_to_markdown.py
// auch in: OpenClaw@gateway2:skills/web-markdown-scraper/scripts/scrape_to_markdown.py
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

const fs = require('fs');
const path = require('path');
const { URL } = require('url');
const { program } = require('commander');
const axios = require('axios');
const cheerio = require('cheerio');
const TurndownService = require('turndown');

function toStr(value) {
    if (value === null || value === undefined) {
        return "";
    }
    if (Buffer.isBuffer(value)) {
        return value.toString('utf-8');
    }
    return String(value);
}

function slugify(text, maxLen = 80) {
    let result = text.replace(/[^\w\s-]/g, '').trim().toLowerCase();
    result = result.replace(/[-\s]+/g, '-');
    result = result.substring(0, maxLen).replace(/-+$/, '');
    return result || "page";
}

function extractHtml(obj) {
    if (obj === null || obj === undefined) {
        return "";
    }
    
    const attrs = ["html", "rawHtml", "content", "markup", "body", "innerHtml"];
    for (const attr of attrs) {
        let value = obj[attr];
        if (typeof value === 'function') {
            try {
                value = value.call(obj);
            } catch (e) {
                value = null;
            }
        }
        const text = toStr(value);
        if (text && text.includes("<") && text.includes(">")) {
            return text;
        }
    }
    
    const text = toStr(obj);
    return (text.includes("<") && text.includes(">")) ? text : "";
}

function extractTitle(html) {
    const match = html.match(/<title[^>]*>(.*?)<\/title>/is);
    if (!match) {
        return "";
    }
    let title = match[1].replace(/<[^>]+>/g, " ");
    title = title.replace(/\s+/g, " ").trim();
    return title;
}

async function fetchPage(url, options = {}) {
    const { timeout = 30000 } = options;
    
    try {
        const response = await axios.get(url, {
            timeout: timeout,
            headers: {
                'User-Agent': 'Mozilla/5.0 (compatible; Scraper/1.0)'
            }
        });
        
        return {
            data: response.data,
            status: response.status,
            statusText: response.statusText,
            headers: response.headers
        };
    } catch (error) {
        throw new Error(`Failed to fetch ${url}: ${error.message}`);
    }
}

function pickMainHtml(pageData, preferredSelector = null) {
    const $ = cheerio.load(pageData);
    const selectors = [];
    
    if (preferredSelector) {
        selectors.push(preferredSelector);
    }
    
    selectors.push(
        "article",
        "main",
        "[role='main']",
        ".post-content",
        ".entry-content",
        ".article-content",
        "body"
    );
    
    for (const selector of selectors) {
        try {
            const element = $(selector).first();
            if (element.length > 0) {
                const html = element.html();
                if (html && html.length >= 120) {
                    return [html, selector];
                }
            }
        } catch (e) {
            // Continue to next selector
        }
    }
    
    return [pageData, null];
}

function htmlToMarkdown(html, preserveLinks = false, bodyWidth = 0) {
    const turndownService = new TurndownService({
        headingStyle: 'atx',
        hr: '---',
        bulletListMarker: '-',
        codeBlockStyle: 'fenced',
        emDelimiter: '*',
        strongDelimiter: '**',
        linkStyle: preserveLinks ? 'inlined' : 'referenced',
        linkReferenceStyle: 'full'
    });
    
    // Remove images
    turndownService.remove('img');
    
    // Handle tables
    turndownService.addRule('table', {
        filter: 'table',
        replacement: function(content) {
            return content;
        }
    });
    
    let markdown = turndownService.turndown(html);
    markdown = markdown.replace(/\n{3,}/g, '\n\n').trim();
    return markdown;
}

function loadUrls(urlArgs = [], urlFile = "") {
    let urls = [...urlArgs];
    
    if (urlFile) {
        const content = fs.readFileSync(urlFile, 'utf-8');
        const lines = content.split('\n');
        for (const line of lines) {
            const trimmed = line.trim();
            if (trimmed && !trimmed.startsWith('#')) {
                urls.push(trimmed);
            }
        }
    }
    
    // Deduplicate URLs
    const seen = new Set();
    const clean = [];
    for (const u of urls) {
        if (!seen.has(u)) {
            clean.push(u);
            seen.add(u);
        }
    }
    
    return clean;
}

function validateUrl(url) {
    try {
        const parsed = new URL(url);
        return ['http:', 'https:'].includes(parsed.protocol) && Boolean(parsed.hostname);
    } catch (e) {
        return false;
    }
}

async function main() {
    program
        .option('--url <urls...>', 'URLs to scrape')
        .option('--url-file <file>', 'File containing URLs to scrape')
        .option('--selector <selector>', 'Preferred CSS selector for content')
        .option('--js', 'Use JavaScript rendering (not supported in this version)')
        .option('--wait-selector <selector>', 'Wait for selector (not supported in this version)')
        .option('--preserve-links', 'Preserve links in markdown output')
        .option('--body-width <width>', 'Body width for markdown output', parseInt)
        .option('--timeout <timeout>', 'Request timeout in seconds', parseInt, 30)
        .option('--output-dir <dir>', 'Output directory', 'outputs')
        .option('--automatch-domain <domain>', 'Automatch domain (not supported in this version)');
    
    program.parse();
    const options = program.opts();
    
    const urls = loadUrls(options.url, options.urlFile);
    if (urls.length === 0) {
        console.log(JSON.stringify({ok: false, error: "No URLs provided"}));
        process.exit(1);
    }
    
    for (const u of urls) {
        if (!validateUrl(u)) {
            console.log(JSON.stringify({ok: false, error: `Invalid URL: ${u}`}));
            process.exit(1);
        }
    }
    
    const outputDir = path.resolve(options.outputDir);
    fs.mkdirSync(outputDir, { recursive: true });
    
    const results = [];
    
    for (const url of urls) {
        const item = {
            url: url,
            ok: false,
            title: "",
            status: null,
            selector_used: null,
            backend: null,
            markdown: "",
            preview: "",
            output_markdown_file: null,
            error: null,
        };
        
        try {
            const page = await fetchPage(url, { timeout: options.timeout * 1000 });
            const [html, selectorUsed] = pickMainHtml(page.data, options.selector || null);
            
            if (!html) {
                throw new Error("No HTML content extracted from page");
            }
            
            const title = extractTitle(html) || new URL(url).hostname;
            const markdown = htmlToMarkdown(
                html,
                options.preserveLinks,
                options.bodyWidth
            );
            
            const filename = slugify(`${new URL(url).hostname}-${title}`) + ".md";
            const mdPath = path.join(outputDir, filename);
            fs.writeFileSync(mdPath, markdown, 'utf-8');
            
            item.ok = true;
            item.title = title;
            item.status = page.status;
            item.selector_used = selectorUsed;
            item.backend = "axios";
            item.markdown = markdown;
            item.preview = markdown.substring(0, 1200);
            item.output_markdown_file = mdPath;
        } catch (e) {
            item.error = e.message;
        }
        
        results.push(item);
    }
    
    const ok = results.some(x => x.ok);
    const indexPath = path.join(outputDir, "index.json");
    const payload = {
        ok: ok,
        count: results.length,
        success_count: results.filter(x => x.ok).length,
        failure_count: results.filter(x => !x.ok).length,
        output_index_file: indexPath,
        results: results
    };
    
    fs.writeFileSync(indexPath, JSON.stringify(payload, null, 2), 'utf-8');
    console.log(JSON.stringify(payload));
}

if (require.main === module) {
    main().catch(err => {
        console.error(err);
        process.exit(1);
    });
}

File: perl5/scrape_to_markdown.pl
#!/usr/bin/perl
# scrape_to_markdown.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:skills/web-markdown-scraper/scripts/scrape_to_markdown.py
# auch in: OpenClaw@gateway2:skills/web-markdown-scraper/scripts/scrape_to_markdown.py
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use utf8;
use JSON;
use URI;
use URI::Escape;
use File::Path qw(make_path);
use File::Basename;
use Getopt::Long;
use LWP::UserAgent;
use HTML::TreeBuilder;
use HTML::FormatText::WithLinks;

# SECURITY MANIFEST:
# Environment variables accessed: none
# External endpoints called: only URLs supplied by the user at runtime via --url / --url-file
# Local files read: --url-file path (if provided by user)
# Local files written: --output-dir/*.md, --output-dir/index.json (if --output-dir provided)
#                      Scrapling automatch SQLite DB (managed by Scrapling, local only)
# Credentials handled: --proxy value (never logged or transmitted beyond the proxy itself)
# Shell injection risk: none (pure Perl, no system calls or shell interpolation)

sub to_str {
    my ($value) = @_;
    return "" unless defined $value;
    if (ref($value) eq 'SCALAR' && ref($value) eq 'GLOB') {
        # Handle filehandle-like objects if needed
        return "";
    }
    return $$value if ref($value) eq 'SCALAR';
    return "$value";
}

sub slugify {
    my ($text, $max_len) = @_;
    $max_len //= 80;
    $text =~ s/[^\w\s-]//g;
    $text =~ s/[-\s]+/-/g;
    $text = lc($text);
    $text = substr($text, 0, $max_len);
    $text =~ s/^-+|-+$//g;
    return $text || "page";
}

sub extract_html {
    my ($obj) = @_;
    return "" unless defined $obj;
    
    my @attrs = qw(html raw_html content markup body inner_html);
    for my $attr (@attrs) {
        if (ref($obj) && $obj->can($attr)) {
            my $value = eval { $obj->$attr() };
            next if $@;
            my $text = to_str($value);
            return $text if $text && $text =~ /</ && $text =~ />/;
        }
    }
    
    my $text = to_str($obj);
    return ($text =~ /</ && $text =~ />/) ? $text : "";
}

sub extract_title {
    my ($html) = @_;
    return "" unless defined $html;
    
    if ($html =~ /<title[^>]*>(.*?)<\/title>/si) {
        my $title = $1;
        $title =~ s/<[^>]+>/ /g;
        $title =~ s/\s+/ /g;
        $title =~ s/^\s+|\s+$//g;
        return $title;
    }
    return "";
}

sub fetch_page {
    my ($url, $js, $wait_selector, $timeout, $automatch_domain) = @_;
    
    my $ua = LWP::UserAgent->new;
    $ua->timeout($timeout // 30);
    my $response = $ua->get($url);
    
    if ($response->is_success) {
        return ($response->decoded_content, "LWP::UserAgent.get");
    } else {
        die "Failed to fetch $url: " . $response->status_line;
    }
}

sub pick_main_html {
    my ($page, $preferred_selector) = @_;
    
    my @selectors = ();
    push @selectors, $preferred_selector if $preferred_selector;
    push @selectors, qw(article main [role='main'] .post-content .entry-content .article-content body);
    
    # Simple implementation without CSS selection
    return (extract_html($page), undef);
}

sub html_to_markdown {
    my ($html, $preserve_links, $body_width) = @_;
    $body_width //= 0;
    
    my $h = HTML::FormatText::WithLinks->new(
        before_link => '',
        after_link => ' [%l]',
        footnote => $preserve_links ? '[%n] %l' : '',
        links_after_each_paragraph => 0,
        display_links => $preserve_links,
        remove_newlines => 1,
        with_emphasis => 1,
    );
    
    my $md = $h->parse($html);
    $md =~ s/\n{3,}/\n\n/g;
    $md =~ s/^\s+|\s+$//g;
    return $md;
}

sub load_urls {
    my ($url_args, $url_file) = @_;
    
    my @urls = @$url_args;
    
    if ($url_file && -f $url_file) {
        open my $fh, '<:encoding(UTF-8)', $url_file or die "Cannot open $url_file: $!";
        while (my $line = <$fh>) {
            chomp $line;
            $line =~ s/^\s+|\s+$//g;
            next if !$line || $line =~ /^#/;
            push @urls, $line;
        }
        close $fh;
    }
    
    my %seen;
    my @clean;
    for my $u (@urls) {
        if (!$seen{$u}) {
            push @clean, $u;
            $seen{$u} = 1;
        }
    }
    
    return @clean;
}

sub validate_url {
    my ($url) = @_;
    my $uri = URI->new($url);
    return ($uri->scheme =~ /^(http|https)$/i) && $uri->host;
}

sub main {
    my @urls;
    my $url_file = "";
    my $selector = "";
    my $js = 0;
    my $wait_selector = "";
    my $preserve_links = 0;
    my $body_width = 0;
    my $timeout = 30;
    my $output_dir = "outputs";
    my $automatch_domain = "";
    
    GetOptions(
        "url=s@" => \@urls,
        "url-file=s" => \$url_file,
        "selector=s" => \$selector,
        "js" => \$js,
        "wait-selector=s" => \$wait_selector,
        "preserve-links" => \$preserve_links,
        "body-width=i" => \$body_width,
        "timeout=i" => \$timeout,
        "output-dir=s" => \$output_dir,
        "automatch-domain=s" => \$automatch_domain,
    ) or die "Error in command line arguments\n";
    
    my @loaded_urls = load_urls(\@urls, $url_file);
    
    if (!@loaded_urls) {
        print encode_json({ok => JSON::false, error => "No URLs provided"}) . "\n";
        exit 1;
    }
    
    for my $u (@loaded_urls) {
        if (!validate_url($u)) {
            print encode_json({ok => JSON::false, error => "Invalid URL: $u"}) . "\n";
            exit 1;
        }
    }
    
    make_path($output_dir) unless -d $output_dir;
    
    my @results;
    
    for my $url (@loaded_urls) {
        my $item = {
            url => $url,
            ok => JSON::false,
            title => "",
            status => undef,
            selector_used => undef,
            backend => undef,
            markdown => "",
            preview => "",
            output_markdown_file => undef,
            error => undef,
        };
        
        eval {
            my ($page, $backend) = fetch_page(
                $url,
                $js,
                $wait_selector || undef,
                $timeout,
                $automatch_domain || undef,
            );
            
            my ($html, $selector_used) = pick_main_html($page, $selector || undef);
            die "No HTML content extracted from page" unless $html;
            
            my $title = extract_title($html) || do {
                my $uri = URI->new($url);
                $uri->host;
            };
            
            my $markdown = html_to_markdown(
                $html,
                $preserve_links,
                $body_width,
            );
            
            my $filename = slugify(URI->new($url)->host . "-$title") . ".md";
            my $md_path = "$output_dir/$filename";
            
            open my $fh, '>:encoding(UTF-8)', $md_path or die "Cannot write to $md_path: $!";
            print $fh $markdown;
            close $fh;
            
            $item->{ok} = JSON::true;
            $item->{title} = $title;
            $item->{status} = 200;  # Simplified
            $item->{selector_used} = $selector_used;
            $item->{backend} = $backend;
            $item->{markdown} = $markdown;
            $item->{preview} = substr($markdown, 0, 1200);
            $item->{output_markdown_file} = $md_path;
        };
        
        if ($@) {
            $item->{error} = "$@";
        }
        
        push @results, $item;
    }
    
    my $ok = grep { $_->{ok} } @results;
    my $index_path = "$output_dir/index.json";
    
    my $payload = {
        ok => $ok ? JSON::true : JSON::false,
        count => scalar(@results),
        success_count => scalar(grep { $_->{ok} } @results),
        failure_count => scalar(grep { !$_->{ok} } @results),
        output_index_file => $index_path,
        results => \@results,
    };
    
    open my $fh, '>:encoding(UTF-8)', $index_path or die "Cannot write to $index_path: $!";
    print $fh JSON->new->pretty->encode($payload);
    close $fh;
    
    print JSON->new->encode($payload) . "\n";
}

main() unless caller;
